src/tlo/methods/enhanced_lifestyle.py

```diff
@@ -825,533 +428,1495 @@ def on_birth(self, mother_id, child_id):
+            eligible_idx = df.loc[df.is_alive & (df.sex == 'F') & ~df.li_is_sexworker & df.age_years.between(15, 49)
+                                  & (df.li_mar_stat != 2)].index
+
+            n_sw = len(df.loc[df.is_alive & df.li_is_sexworker].index)
+            target_n_sw = int(np.round(len(df.loc[df.is_alive & (df.sex == 'F') & df.age_years.between(15, 49)].index)
+                                       * p["proportion_female_sex_workers"]
+                                       )
+                              )
+            deficit = target_n_sw - n_sw
+            if deficit > 0:
+                if deficit < len(eligible_idx):
+                    # randomly select women to start sex work:
+                    will_start_sw_idx = rng.choice(eligible_idx, deficit, replace=False)
+                else:
+                    # select all eligible women to start sex work:
+                    will_start_sw_idx = eligible_idx
+                # make is_sexworker for selected women:
+                f_sex_worker.loc[will_start_sw_idx] = True
+
+            # return updated female sexworker series
+            return f_sex_worker
+
+        # return female sex worker linear model
+        f_sex_worker_lm = LinearModel.custom(predict_female_sex_workers, parameters=self.params)
+        return f_sex_worker_lm
+
+    def male_circumsision_property_linear_model(self) -> LinearModel:
+        """A function to create linear model for initialising the male circumcision property. Here, we are using
+        custom linear model """
+
+        def handle_male_circumcision_prop(self, df, rng=None, **externals) -> pd.Series:
+            """ # determine the proportion of men that are circumcised at initiation
+            # NB. this is determined with respect to any characteristics (eg. ethnicity or religion)
+            Steps:
+                1.  accept a population dataframe from the predict method of a custom linear model
+                2.  create a new series using `li_is_circ` column data in the population dataframe
+                3.  update the series using different probabilities
+                4.  return the updated series
+
+            :param self: a reference to the calling custom linear model
+            :param df: population dataframe
+            :param rng: random number generator
+            :param  externals a dict containing any other variables passed on to this function """
+
+            # get module parameters
+            p = self.parameters
+
+            # create a new male circumcision series
+            male_circ = pd.Series(data=False, index=df.index, dtype=bool)
+
+            # select a population of men to be circumcised
+            men = df.loc[df.is_alive & (df.sex == 'M')]
+            will_be_circ = rng.rand(len(men)) < p['proportion_of_men_circumcised_at_initiation']
+            male_circ.loc[men[will_be_circ].index] = True
```

Comment:
```suggestion
            is_alive_and_male = df.is_alive & (df.sex == 'M')
            will_be_circ = rng.rand(is_alive_and_male.sum()) < p['proportion_of_men_circumcised_at_initiation']
            male_circ.loc[is_alive_and_male & will_be_circ] = True
```
Avoids the unnecessary `df.loc` indexing operation

Reply:
This implementation is failing here `male_circ.loc[is_alive_and_male & will_be_circ] = True`

```
tests\test_enhanced_lifestyle.py:58 (test_properties_and_dtypes[83563095832589325021])
test_enhanced_lifestyle.py:60: in test_properties_and_dtypes
    simulation.make_initial_population(n=popsize)
..\src\tlo\simulation.py:189: in make_initial_population
    module.initialise_population(self.population)
..\src\tlo\methods\enhanced_lifestyle.py:374: in initialise_population
    self.models.initialise_all_properties(df)
..\src\tlo\methods\enhanced_lifestyle.py:653: in initialise_all_properties
    df.loc[df.is_alive, _property_name] = _model['init'].predict(
..\src\tlo\methods\enhanced_lifestyle.py:1169: in handle_male_circumcision_prop
    male_circ.loc[is_alive_and_male & will_be_circ] = True
..\..\..\anaconda3\envs\tlo\lib\site-packages\pandas\core\ops\common.py:81: in new_method
    return method(self, other)
..\..\..\anaconda3\envs\tlo\lib\site-packages\pandas\core\arraylike.py:70: in __and__
    return self._logical_method(other, operator.and_)
..\..\..\anaconda3\envs\tlo\lib\site-packages\pandas\core\series.py:6107: in _logical_method
    res_values = ops.logical_op(lvalues, rvalues, op)
..\..\..\anaconda3\envs\tlo\lib\site-packages\pandas\core\ops\array_ops.py:401: in logical_op
    res_values = na_logical_op(lvalues, rvalues, op)
..\..\..\anaconda3\envs\tlo\lib\site-packages\pandas\core\ops\array_ops.py:311: in na_logical_op
    result = op(x, y)
E   ValueError: operands could not be broadcast together with shapes (10000,) (4946,)
```

Reply:
deferred